src/storage/database/knex.ts

```diff
@@ -0,0 +1,567 @@
+      return knex
+        .from<Obj>('objects')
+        .update({
+          metadata,
+        })
+        .where({ bucket_id: bucketId, name: objectName })
+        .returning('*')
+    })
+
+    return object
+  }
+
+  async updateObjectOwner(bucketId: string, objectName: string, owner?: string) {
+    const [object] = await this.runQuery('UpdateObjectOwner', (knex) => {
+      return knex
+        .from<Obj>('objects')
+        .update({
+          last_accessed_at: new Date().toISOString(),
+          owner,
+        })
+        .returning('*')
+        .where({ bucket_id: bucketId, name: objectName })
+    })
+
+    if (!object) {
+      throw new DBError('Object not found', 404, 'not_found', undefined, {
+        bucketId,
+      })
+    }
+
+    return object
+  }
+
+  async findObject(
+    bucketId: string,
+    objectName: string,
+    columns = 'id',
+    filters?: FindObjectFilters
+  ) {
+    const object = await this.runQuery('FindObject', (knex) => {
+      const query = knex.from<Obj>('objects').select(columns.split(',')).where({
+        name: objectName,
+        bucket_id: bucketId,
+      })
+
+      if (filters?.forUpdate) {
+        query.forUpdate()
+      }
+
+      if (filters?.forShare) {
+        query.forShare()
+      }
+
+      if (filters?.forKeyShare) {
+        query.forKeyShare()
+      }
+
+      if (filters?.noWait) {
+        query.noWait()
+      }
+
+      return query.first() as Promise<Obj | undefined>
+    })
+
+    if (!object && !filters?.dontErrorOnEmpty) {
+      throw new DBError('Object not found', 404, 'not_found', undefined, {
+        bucketId,
+      })
+    }
+
+    return object as typeof filters extends FindObjectFilters
+      ? typeof filters['dontErrorOnEmpty'] extends true
+        ? Obj | undefined
+        : Obj
+      : Obj
+  }
+
+  async findObjects(bucketId: string, objectNames: string[], columns = 'id') {
+    const objects = await this.runQuery('FindObjects', (knex) => {
+      return knex
+        .from<Obj>('objects')
+        .select(columns)
+        .where('bucket_id', bucketId)
+        .whereIn('name', objectNames)
+    })
+
+    return objects
+  }
+
+  async mustLockObject(bucketId: string, objectName: string, version?: string) {
+    return this.runQuery('MustLockObject', async (knex) => {
+      const hash = hashStringToInt(`${bucketId}/${objectName}${version ? `/${version}` : ''}`)
+      const result = await knex.raw<any>(`SELECT pg_try_advisory_xact_lock(?);`, [hash])
+      const lockAcquired = result.rows.shift()?.pg_try_advisory_xact_lock || false
+
+      if (!lockAcquired) {
+        throw new DBError('resource_locked', 409, 'Resource is locked')
+      }
+
+      return true
+    })
+  }
+
+  async waitObjectLock(bucketId: string, objectName: string, version?: string) {
+    return this.runQuery('WaitObjectLock', async (knex) => {
+      const hash = hashStringToInt(`${bucketId}/${objectName}${version ? `/${version}` : ''}`)
```

Comment:
should the lock include the tenantID?

Reply:
in theory we are already on the tenant db at this point